presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
-
-        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
-        checkArgument(
-                tableScans.size() == 1,
-                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
-                fragment.getId(),
-                tableScans.size());
-
-        TableScanNode tableScan = getOnlyElement(tableScans);
-
-        List<ScheduledSplit> splits = getSplits(session, tableScan);
-        shuffle(splits);
-        int initialPartitionCount = getSparkInitialPartitionCount(session);
-        int numTasks = Math.min(splits.size(), initialPartitionCount);
-        if (numTasks == 0) {
-            return JavaPairRDD.fromJavaRDD(sparkContext.emptyRDD());
+        ListMultimap<Integer, TaskSource> taskSourcesMap = ArrayListMultimap.create();
+        for (TableScanNode tableScan : tableScans) {
+            List<ScheduledSplit> scheduledSplits = getSplits(session, tableScan);
+            shuffle(scheduledSplits);
+            SetMultimap<Integer, ScheduledSplit> assignedSplits = assignSplitsToTasks(session, partitioning, scheduledSplits);
+            asMap(assignedSplits).forEach((partitionId, splits) ->
+                    taskSourcesMap.put(partitionId, new TaskSource(tableScan.getId(), splits, true)));
         }
 
-        List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(splits, numTasks);
-
-        // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
-        splits = null;
-
-        ImmutableList.Builder<SerializedPrestoSparkTaskDescriptor> serializedTaskDescriptors = ImmutableList.builder();
-        for (int i = 0; i < assignedSplits.size(); i++) {
-            List<ScheduledSplit> splitBatch = assignedSplits.get(i);
-            PrestoSparkTaskDescriptor taskDescriptor = createSourceTaskDescriptor(session, tableWriteInfo, fragment, splitBatch);
-            // TODO: consider more efficient serialization or apply compression to save precious memory on the Driver
-            byte[] jsonSerializedTaskDescriptor = taskDescriptorJsonCodec.toJsonBytes(taskDescriptor);
-            serializedTaskDescriptors.add(new SerializedPrestoSparkTaskDescriptor(jsonSerializedTaskDescriptor));
-            // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
-            assignedSplits.set(i, null);
+        IntStream partitions = expectedNumberOfPartitions
+                .map(integer -> IntStream.range(0, integer))
+                .orElseGet(() -> new ArrayList<>(taskSourcesMap.keySet()).stream().mapToInt(Integer::intValue));
+
+        List<List<SerializedTaskSource>> partitionedTaskSources = new ArrayList<>();
+        partitions.forEach(partition -> {
+            List<TaskSource> taskSources = taskSourcesMap.removeAll(partition);
```

Comment:
Add a comment explain why eagerly remove the task sources

Reply:
Added a comment